### PR TITLE
Change /modlog to match on IDs

### PR DESCRIPTION
**Pull request by Hidden50**
The keyword search doesn't always find all relevant entries, because
users have different ways of formatting their username.

With this change, there will be two ways to do a keyword search:
/modlog n - ignores non-word characters (\W*) in between letters. Edit: For example, /modlog ausername will also bring up matches like "a username" or "a user-name".
/modlog "n" - strictly matches only the given formatting of n. (This is
how /modlog worked before.)

I also changed /modlog [number] to sort matches newest-to-oldest, like
the keyword search does.

These changes don't apply to PS servers running on windows. modlog will
behave the same as before for them

### CODE REVIEW

File: commands.js [hunk 1]
```diff
@@ -1760,15 +1760,39 @@ exports.commands = {
 		if (isWin) {
 			command = path.normalize(__dirname + '/lib/winmodlog') + ' tail ' + lines + ' ' + filename;
 		} else {
-			command = 'tail -' + lines + ' ' + filename;
+			command = 'tail -' + lines + ' ' + filename + ' | tac';
```
Comment (Hidden50) [inline commands.js]: if somebody wants to show me how to do this (reverse tail output line order) on a windows PS server, be my guest n_n

Code for tail on windows PS: https://github.com/Zarel/Pokemon-Showdown/blob/master/lib/winmodlog.cmd#L15

File: commands.js [hunk 2]
```diff
@@ -1760,15 +1760,39 @@ exports.commands = {
...
+					command = path.normalize(__dirname + '/lib/winmodlog') + ' ws ' + grepLimit + ' "' + searchString.replace(/%/g, "%%").replace(/([\^"&<>\|])/g, "^$1") + '" ' + filename;
+				} else {
+					// doesn't happen. ID search with RegEx isn't implemented for windows yet (feel free to add it to winmodlog.cmd and call it from here)
+				}
 			} else {
-				command = "awk '{print NR,$0}' " + filename + " | sort -nr | cut -d' ' -f2- | grep -m" + grepLimit + " -i '" + target.replace(/\\/g, '\\\\\\\\').replace(/["'`]/g, '\'\\$&\'').replace(/[\{\}\[\]\(\)\$\^\.\?\+\-\*]/g, '[$&]') + "'";
+				if (strictMatch) {
+					command = "awk '{print NR,$0}' " + filename + " | sort -nr | cut -d' ' -f2- | grep -m" + grepLimit + " -i '" + searchString.replace(/\\/g, '\\\\\\\\').replace(/["'`]/g, '\'\\$&\'').replace(/[\{\}\[\]\(\)\$\^\.\?\+\-\*]/g, '[$&]') + "'";
+				} else {
+					command = "awk '{print NR,$0}' " + filename + " | sort -nr | cut -d' ' -f2- | grep -m" + grepLimit + " -Ei '" + searchString + "'";
```
Comment (Hidden50) [inline commands.js]: if somebody wants to show me how to do this (do a RegEx search) on a windows PS server, again, be my guest.

Code for regular word-search on windows: https://github.com/Zarel/Pokemon-Showdown/blob/master/lib/winmodlog.cmd#L30

**Comment by Zarel** [comment]
@Joimer?

**Comment by Joimer** [comment]
Using ``tac`` is a very good idea to get them from current to older. Other than that, I'm afraid such a loose search can give many unwanted results on a modlog as large as the main server's, despite the boundaries applied?

**Comment by Hidden50** [comment]
The main sources of collision should be names that are combinations of whole words (like orda-y, which would match on a line containing "night or day" because it's the words "or" and "day" combined).

You can still do /modlog "orda-y" to get the old results though, so I don't really see the problem.
I also don't think those names are too common.

Rather it's users commonly putting random symbols into their name that make us miss results. (i.e. do /modlog [overused], ▲ - all of those are results that don't normally show when you search for the username). Or sometimes a name is logged as its ID, which removes spaces and doesn't show up in the search.

You can do some test cases to check for unwanted results.
(Maybe pull the changes to a test server[not windows] and copy part of the main server's modlog onto it.)

**Comment by Zarel** [comment]
I was more tagging Joim for Windows support. I think this approach is fine.

**Comment by Joimer** [comment]
I just use Windows for games / C#, so I don't know how to implement this in batch :/

**Comment by Zarel** [comment]
Do you remember who made our existing Windows support? @Slayer95?

**Comment by Joimer** [comment]
iirc the Windows support was actually made by a contributor that hadn't done much besides that, he wasn't a usual one.

**Comment by Zarel** [comment]
Oh well. I'm merging unless someone objects.

**Comment by Hidden50** [comment]
I asked one of my contacts who might know a thing or two about batch files. Might want to wait until I hear back.

For me, the batch code is very hard to read.

**Comment by Slayer95** [comment]
It was @Freigeist

File: commands.js [hunk 3]
```diff
@@ -1760,15 +1760,39 @@ exports.commands = {
...
+			strictMatch = true; // search for a 1:1 match?
+
+			if (searchString.match(/^["'].+["']$/)) {
+				searchString = searchString.substring(1, searchString.length - 1);
+			} else if (isWin) {  // ID search with RegEx isn't implemented for windows yet (feel free to add it to winmodlog.cmd)
+				target = '"' + target + '"';  // add quotes to target so the caller knows they are getting a strict match
+			} else {
+				// search for ID: allow any number of non-word characters (\W*) in between the letters we have to match.
+				// i.e. if searching for "myUsername", also match on "My User-Name".
+				// note that this doesn't really add a lot of unwanted results, since we use \b..\b
```
Comment (Zarel) [inline commands.js]: wait, shouldn't we only do an ID search if we input an ID?

Like, only if `target === toId(target)`?

Comment (Hidden50) [inline commands.js]: ah, I think it's more convenient to be able to tab complete (which doesn't give you IDs, you'd have to manually edit it).
I think doing /modlog "a username" is fine for not doing an ID search, this way you can tab complete for both versions.

Comment (Zarel) [inline commands.js]: I suppose. We can change it based on feedback, anyway.

**Pull request merged**